Adds PublisherStore for greater flexibility

GabeVillalobos opened the pull request:
PublisherStore trait adds the ability for developers to add their own publisher backing store with their own thread-safe implementations/concurrency requirements.

GabeVillalobos committed:
Adds PublisherStore for greater flexibility

PublisherStore trait adds the ability for developers to add their own
publisher backing store with their own thread-safe
implementations/concurrency requirements.

GabeVillalobos commented:
The `PublisherStore` trait could use some work/better definition, but figured I'd get the initial implementation up for review at least

andrewjj20 reviewed (commented):
This looks good. It is coming to mind that we don't have great testing on the cleanup code. This is going to need some manual testing or a commit to add that the testing for that. This would be a great candidate for an integration test as well.

GabeVillalobos commented:
Yeah I'll try to add a test for that. Ideally testing each of the cleanup branches would be the goal, but I may need a hand getting each of those conditions to occur

GabeVillalobos committed:
Renames some publisher defs for clarity

vec_publisher_store.find_publishers now does a 'contains' check for
matching publisher names

GabeVillalobos committed:
Adds test for publisher cleanup & fixes cleanup logic

GabeVillalobos commented on .github/workflows/rust.yml:
```diff
@@ -27,6 +27,8 @@ jobs:
       run: cargo clippy --workspace --exclude tests
     - name: Run tests
-      run: cargo test --verbose
+      run: cargo test --verbose -- --test-threads=1
```
Needed `test-threads=1` to prevent races around the meetme service when tests are kicked off in parallel. In the future, we could just select an unallocated port per test.

andrewjj20 replied:
probably need to open up an issue for that.

GabeVillalobos commented on tests/common/find_service_setup.rs:
```diff
@@ -39,6 +39,10 @@ impl FindService {
         let bind = "127.0.0.1:8080";
 
         let _proc = tokio::process::Command::new(path)
+            .arg("-s")
+            .arg("5")
+            .arg("-t")
+            .arg("5")
```
Wasn't sure if this was a reasonable global setting, but we likely want to make this configurable on a per-test basis in the future.

GabeVillalobos commented on tests/publisher_subscriber.rs:
```diff
@@ -134,3 +133,66 @@ async fn publisher_subscriber_basics() {
+    };
+
+    log::debug!("Creating publisher");
+    let mut _publisher = post::publisher::Publisher::from_description(desc.clone(), client.clone())
```
This line has weird drop behavior.

yaahc commented on src/find_service/server.rs:
```diff
@@ -15,73 +12,54 @@ fn convert_system_time_error(time_error: time::SystemTimeError) -> io::Error {
+pub trait PublisherStore: Send + Sync + Clone + 'static {
+    fn insert_publisher(&self, publisher_name: &String, registration: proto::Registration);
+    fn remove_publisher(&self, publisher_name: &String) -> Result<(), PublisherNotFoundError>;
+    fn remove_publishers(&self, publisher_names: &Vec<String>);
```
This should probably provide a default implementation
```suggestion
    fn remove_publishers(&self, publisher_names: &Vec<String>) {
        for publisher in publisher_names {
            self.remove_publisher(publisher);
        }
    }
```

andrewjj20 replied:
Another debate I have is `&Vec<T>` vs `&[T]`. it seems like there is not much difference for our purposes. and `&[T]` has less baggage.

yaahc commented on src/find_service/server.rs:
```diff
@@ -15,73 +12,54 @@ fn convert_system_time_error(time_error: time::SystemTimeError) -> io::Error {
+}
+
+pub trait PublisherStore: Send + Sync + Clone + 'static {
+    fn insert_publisher(&self, publisher_name: &String, registration: proto::Registration);
```
Prefer `&str` over `&String` because it's more flexible and deref coercion will automatically turn a `&String` into a `&str`.

```suggestion
    fn insert_publisher(&self, publisher_name: &str, registration: proto::Registration);
```

yaahc commented on src/find_service/server.rs:
```diff
@@ -15,73 +12,54 @@ fn convert_system_time_error(time_error: time::SystemTimeError) -> io::Error {
+pub trait PublisherStore: Send + Sync + Clone + 'static {
+    fn insert_publisher(&self, publisher_name: &String, registration: proto::Registration);
+    fn remove_publisher(&self, publisher_name: &String) -> Result<(), PublisherNotFoundError>;
+    fn remove_publishers(&self, publisher_names: &Vec<String>);
```
same as with strings, prefer slice references to vec references
```suggestion
    fn remove_publishers(&self, publisher_names: &[String]);
```

You might want to go even further and also allow the `String` type to be flexible:

```rust
    fn remove_publishers<T>(&self, publisher_names: &[T])
    where
        T: Borrow<str>;
```

yaahc replied:
generics might be bad here though, depending on if you need object safety.

GabeVillalobos replied:
Not sure it's worth making this generic yet, updated this to use slice refs as suggested though.

yaahc reviewed (commented):
didn't do a full review, just looked at the trait, lmk if you want a more thorough review.

andrewjj20 commented on src/find_service/vec_publisher_store.rs:
```diff
@@ -0,0 +1,160 @@
+    publishers_map: PublisherStoreMap,
+}
+
+impl PublisherStore for VecPublisherStore {
```
I remember bringing this up in person; just duplicating this here. Can't this be an impl for `PublisherStoreMap` instead of `VecPublisherStore`. Also, wouldn't this be `HashPublisherStore`?

GabeVillalobos replied:
Good call, renamed this to `HashPublisherStore`, also will be testing the  type impl idea now

andrewjj20 reviewed (commented):
wanted to get these comments out before I made another pass.

GabeVillalobos committed:
Renames vec_publisher store, updates PublisherStore trait

GabeVillalobos committed:
Uses type alias for HashMapPublisherStore
